Appext/template generator docs (#10000)
mendix/docs: content/en/docs/apidocs-mxsdk/apidocs/studio-pro-11/extensibility-api/web/get-started.md

## Introduction

Studio Pro extensions can be developed using TypeScript and use standard web development technologies to extend the Studio Pro development environment. This guide shows you how to set up a basic development environment for building an extension using the web extensibility API.

For more detailed information, see the [Mendix Studio Pro Web Extensibility API reference documentation](http://apidocs.rnd.mendix.com/11/extensions-api/index.html).

### Prerequisites

You will need the following prerequisites:

> * [Mendix Studio Pro](https://marketplace.mendix.com/link/studiopro) version 11.2.0 or higher. 

Reviewer: Does this really not work in previous Mx11 versions?

Author: We introduced a breaking change so yes. We also are using our new template generator and this significantly improves the development experience so we thought that encouraging users to use the latest version will help them have the best experience

### Create a Test App

> Create a new app using the **Blank Web App** template. Make sure to take note of the application directory (more precisely, the application *.mpr* file) where the application is stored on disk. 

Reviewer: I don't think we need to do the "Make sure…" bit - the bit in the info box will help customers find the mpr very easily.
```suggestion
Create a new app using the **Blank Web App** template.
```

{{% alert color="info" %}}

> You can also open the application directory by clicking the **App** menu > **Show App Directory in Explorer** (or **Show App Directory in Finder**) in Studio Pro.

Reviewer: ```suggestion
When you need it, you can open the application directory containing the application `.mpr` file, by clicking the **App** menu > **Show App Directory in Explorer** (or **Show App Directory in Finder**) in Studio Pro.
```

{{% /alert %}}

### Creating the Extension

To accelerate your extension development, we provide an extension generator that creates a customizable sample extension.

> To use the generator, navigate to your desired source code directory and run the command `npm create @mendix/extension`. You may be prompted by `npm` to grant permission to install the generator. After installation, you will be guided through a series of questions to help configure your extension.

Reviewer: What is "your desired source code directory"? Does it mean the application directory containing the `.mpr` file? Or is it anywhere you want to work on the code and doesn't need to be related to the mpr at all?

Author: It is pretty standard term for when you are writing code. It is the directory where your source code is located and It does not have anything to do with the app directory. The wizard also draws this distinction by asking you for both the source code directory and the app directory

> You will be asked to select the programming language (TypeScript is used in our tutorials), specify the extension name, and choose whether to use React for the extension’s UI. The next two questions, while optional, are highly recommended, as they enable direct debugging and deployment from Visual Studio Code. These questions request the path to the Studio Pro executable and the application `.mpr` package. The final question allows you to select the Studio Pro version; be sure to choose version 11.

Reviewer: Make this a list of the options?

{{% alert color="info" %}}

> The Studio Pro executable is typically located at `C:\Program Files\Mendix\<version>\modeler\studiopro.exe`. To find the exact path, follow these steps:

Reviewer: Do we want to add the Mac location as well? We mention "Show App Directory in Finder", above?

Reviewer 2: @MarkvanMents Do you know how the Mac location is structured? I'm unsure but can add it if you can shaer the formatting.

Reviewer: Don't know much about Macs. Could just say that this typical location is for Windows machines.

If you want to add the alternatives for Mac, perhaps Nicoleta or Olu can help since they work on Macs?

1. Launch Studio Pro.
2. Right-click its taskbar icon, then right-click `Mendix Studio Pro 11.2.0` (your version may differ).
3. Select **Properties**. The **Target** field displays the executable path.
{{% /alert %}}

Once you have completed the setup, a new directory named after your extension will be created,
containing the source code of the extension.

### Exploring the Created Extension

In the following example, the name of your extension is `myextension` and you are exploring it using
Visual Studio Code.

As a first step your extension will have to get an instance of the Studio Pro api. From the Explorer window, navigate to src/main/index.ts and select it to open the file.

Reading through the source code you should see the following:

1. Line 6 gets an instance of the Studio Pro api by calling getStudioProApi
    ```typescript
    export const component: IComponent = {
        async loaded(componentContext) {

>             const studioPro = getStudioProApi(componentContext);

Reviewer: I'm not sure it is a good idea to explicitly refer to line numbers. They may change in the future?
If we are going to use line numbers, we could add them to the code (but I don't really recommend having line number references at all)
```suggestion
1. Line 6 gets an instance of the Studio Pro api by calling getStudioProApi
    ```typescript {linenos=true linenos=table hl_lines=[3] linenostart=4}
    export const component: IComponent = {
        async loaded(componentContext) {
            const studioPro = getStudioProApi(componentContext);
```

Reviewer 2: Quite a few of the other APIs refer to specific lines, so removing them here would require changes to the other docs with similar wording.

Reviewer 2: @MarcellvanRooyen Thoughts about the line numbers?

2. Line 7 adds a menu

    ```typescript
    await studioPro.ui.extensionsMenu.add({
        menuId: "myextension.MainMenu",
        caption: "MyExtension Menu",
        subMenus: [
            { menuId: "myextension.ShowMenu", caption: "Show tab" },
        ],
    });
    ```

3. Line 16 opens a tab

    ```typescript
    // Open a tab when the menu item is clicked
    studioPro.ui.extensionsMenu.addEventListener(
        "menuItemActivated",
        (args) => {
            if (args.menuId === "myextension.ShowMenu") {
                studioPro.ui.tabs.open(
                    {
                        title: "MyExtension Tab"
                    },
                    {
                        componentName: "extension/myextension",
                        uiEntrypoint: "tab",
                    }
                );
            }
        }
    );
    ```

> 4. If you navigate to `build-extension.mjs`, you can choose the directory to which the extension will be installed to after being built in line 6:

Reviewer: Again, is it a good idea to refer to line numbers which may change (and be different for different SP versions?)

```suggestion
4. If you navigate to `build-extension.mjs`, you can choose the directory to which the extension will be installed to after being built by changing line 6:
```

     ```typescript
     const appDir = "C:\\TestApps\\AppTestExtensions"
     ```

> 5. The file `.vscode\launch.json` specifies the launch configuration and enables debugging. Lines 8-9 specify how Studio Pro will be run:

Reviewer: You could say
```suggestion
5. The file `.vscode\launch.json` specifies the launch configuration and enables debugging. The following lines specify how Studio Pro will be run:
```
to avoid line number issues in the future.

     
     ```json
     "runtimeExecutable": "C:\\Program Files\\Mendix\\11.2.0\\modeler\\studiopro.exe",
     "runtimeArgs": ["C:\\TestApps\\AppTestExtensions\\AppTestExtensions.mpr", "--enable-extension-development", "--enable-web-extensions"],

>      ```

Reviewer: Would it be a good idea to add ellipses to these code snippets to indicate that we know they are not valid JSON unless incorporated in the right place - or would that mess up the copy/paste functionality?
```suggestion
     ```json
     …
     "runtimeExecutable": "C:\\Program Files\\Mendix\\11.2.0\\modeler\\studiopro.exe",
     "runtimeArgs": ["C:\\TestApps\\AppTestExtensions\\AppTestExtensions.mpr", "--enable-extension-development", "--enable-web-extensions"],
     …
     ```
```

When you install the extension, you will see a new menu item within Studio Pro.

> ### Building and Debugging the Extension

Reviewer: ```suggestion
### Building, Installing, and Debugging the Extension
```

The following steps occur within Visual Studio Code:

1. Select **File** > **Open Folder**.
2. Navigate to the folder where you created your extension.
3. Click **Select Folder**.
4. Select **Yes** if you are asked whether you trust this folder.
5. Open a Terminal from the top menu by clicking **Terminal** > **New Terminal**.
6. From the Terminal, type `npm install`. This installs all dependencies for the extension.
7. Build your extension using the command `npm run build` in the terminal. If you provided the path to `.mpr` file in the previous step, this will install the extension into the application directory.

> If the last two questions of the extension generator were answered and you have built and installed the extension, you can debug it by following the steps below:

Reviewer: How do you do it if you didn't provide the `.mpr` location?

Reviewer 2: @MarcellvanRooyen

1. Open the extension source code in Visual Studio Code and set breakpoints.
2. Select **Run and Debug** from the side panel.
3. Click the play button on the top of the panel (or press F5).

> This will run Studio Pro in the extension development mode and open the configured application. You will see a new `Extensions` item in the top menu.

Reviewer: ```suggestion
This will run Studio Pro in extension development mode and open the configured application. You will see a new `Extensions` item in the top menu.
```

### Create a Test App

Create a new app using the **Blank Web App** template.

{{% alert color="info" %}}
You can also open the application directory containing the application `.mpr` file by clicking the **App** menu > **Show App Directory in Explorer** (or **Show App Directory in Finder**) in Studio Pro.
{{% /alert %}}

### Creating the Extension

To accelerate your extension development, we provide an extension generator that creates a customizable sample extension.

To use the generator, navigate to your desired source code directory and run the command `npm create @mendix/extension`. You may be prompted by `npm` to grant permission to install the generator. After installation, you will be guided through a series of questions to help configure your extension.

You will be asked the following:

* Select the programming language (TypeScript is used in our tutorials)
* Specify the extension name
* Choose if you  will use React for the extension’s UI
  

> The next two questions, while optional, are highly recommended, as they enable direct debugging and deployment from Visual Studio Code. These questions request the path to the Studio Pro executable and the application `.mpr` package. The final question allows you to select the Studio Pro version; be sure to choose version 11.

Reviewer 2: What are the two optional questions? I'd like to add them (as well as the final question about selecting the SP version) to the list above so all the setup questions are listed together.

Author: "These questions request the path to the Studio Pro executable and the application `.mpr` package." as stated.
 Do you prefer if we make it bullet points?

Reviewer 2: I would like to add them to the list above, preferably as separate bullets and mark them as optional, but recommended. How are the questions worded in SP? Does the user choose yes/no?

Author: No these questions request you to supply the path. The question highlights that it is optional by adding a (optional at the end of the question. If not supplied it defaults to the current directory so it is pretty much inline with the npm tool itself. Extension developers already have to use that tool so we try to align it closely with that experience. I will add them as separate bullet points and explain that the user supplies the path

Reviewer 2: Thanks!

Author: updated to bullet points and slightly reworded things to read a bit better with the new structure
